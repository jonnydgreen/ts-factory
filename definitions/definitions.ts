import { BlockInput } from './block/block.type.ts';
import { FunctionDeclarationInput } from './function-declaration/function-declaration.type.ts';
import { IdentifierInput } from './identifier/identifier.type.ts';
import { KeywordTypeNodeInput } from './keyword-type-node/keyword-type-node.type.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { SourceFileInput } from './source-file/source-file.type.ts';

// Definition
export type Definition =
  | SourceFileInput
  | StatementInput
  | ExpressionInput
  | TypeNodeInput
  | ModifierInput;

// Member Name
export type MemberNameInput = IdentifierInput;

// Statement Inputs
export type StatementInput = FunctionDeclarationInput | BlockInput;

// Expression Inputs
export type ExpressionInput = IdentifierInput;

// Type Node Inputs
export type TypeNodeInput = KeywordTypeNodeInput;

// Input
export type Input = SourceFileInput;
