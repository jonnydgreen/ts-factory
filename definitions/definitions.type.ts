import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type {
  ExpressionInput,
  MemberNameInput,
  PropertyNameInput,
  StatementInput,
  TypeElementInput,
  TypeNodeInput,
} from './definitions.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { TokenInput } from './token/token.type.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  T extends string | ts.PropertyName ? string | PropertyNameInput :
  T extends ts.Modifier ? ModifierInput :
  T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.TypeElement ? TypeElementInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  T extends ts.Token<ts.SyntaxKind> ? TokenInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;

export type PickDefinitionFields<T> = Omit<
  T,
  '__instructions' | 'leadingTrivia' | 'kind'
>;

export type DefinitionFields<T> = keyof PickDefinitionFields<T>;
