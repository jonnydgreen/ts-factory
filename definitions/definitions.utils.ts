import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type { ExpressionInput, MemberNameInput, StatementInput } from './definitions.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  // T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;
