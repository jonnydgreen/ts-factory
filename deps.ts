// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
