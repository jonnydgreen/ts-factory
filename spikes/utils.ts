export function assertNode<TNode>(node: unknown): asserts node is TNode {
}
