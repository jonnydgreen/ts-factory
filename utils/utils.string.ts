export class StringUtils {
  /**
   * Upper cases the first letter of an input string.
   */
  static upperFirst(input: string): string {
    return `${input[0]?.toUpperCase() ?? ''}${input.slice(1)}`;
  }
}
