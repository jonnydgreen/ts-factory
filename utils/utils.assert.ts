import { Definition } from '../definitions/definitions.ts';
import { ts, tsm } from '../deps.ts';
import { ItemOrArray } from '../instructions/instructions.type.ts';
import { GuardedType, Maybe } from '../types.ts';

export class AssertionError extends Error {
  override name = 'AssertionError';
  constructor(message: string) {
    super(message);
  }
}

export function assertNever(input: never): never {
  throw new AssertionError(`Input ${input} not supported`);
}

export function assertDefinitionKind(
  definition: Definition,
  kind: tsm.ts.SyntaxKind,
): asserts definition is Extract<Definition, { kind: typeof kind }> {
  if (definition.kind !== kind) {
    throw new AssertionError(
      `Definition of kind '${
        tsm.ts.SyntaxKind[definition.kind]
      }' does not match expected Node kind '${tsm.SyntaxKind[kind]}'`,
    );
  }
}

export function assertDefined<T>(input: Maybe<T>): asserts input is T {
  if (typeof input === 'undefined') {
    throw new AssertionError('Input is not defined');
  }
}

export function assertTSNodeType<
  TNode extends ts.Node,
  // deno-lint-ignore no-explicit-any
  TPredicate extends (node: any) => node is any,
>(
  node: Maybe<TNode>,
  predicate: TPredicate,
): asserts node is GuardedType<TPredicate> {
  if (typeof node === 'undefined') {
    throw new AssertionError(`Node is not defined`);
  }

  const kind = node.kind;
  if (!predicate(node)) {
    throw new AssertionError(`Invalid Node of kind ${ts.SyntaxKind[kind]}`);
  }
}

export function assertTSMNodeKind<TKind extends ts.SyntaxKind>(
  node: tsm.Node,
  kind: TKind,
): asserts node is tsm.KindToNodeMappings[TKind] {
  const nodeKind = node?.getKind();
  if (!node?.isKind?.(kind)) {
    throw new AssertionError(
      `Invalid Node of kind ${ts.SyntaxKind[nodeKind]}; expected ${ts.SyntaxKind[kind]}`,
    );
  }
}

export function isDefined<T>(
  nodeOrNodes: Maybe<ItemOrArray<T>>,
): nodeOrNodes is ItemOrArray<T> {
  if (!nodeOrNodes) {
    return false;
  }

  if (Array.isArray(nodeOrNodes) && nodeOrNodes.length === 0) {
    return false;
  }

  return true;
}

export function assertArray<T>(input: Maybe<T> | Maybe<T>[]): asserts input is T[] {
  if (!Array.isArray(input)) {
    throw new AssertionError(`Input is not of type array: ${typeof input}`);
  }
}

export function assertNotArray<T>(input: T | T[]): asserts input is T {
  if (Array.isArray(input)) {
    throw new AssertionError(`Input is of type array when it should not`);
  }
}
