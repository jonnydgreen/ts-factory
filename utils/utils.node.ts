import { ts, tsm } from '../deps.ts';
import { Maybe } from '../types.ts';

export function shouldPrintNode(node?: ts.Node): Maybe<string> {
  if (node) {
    return printNode(node);
  }
}

export function printNode(node: ts.Node): string {
  return tsm.printNode(node, { removeComments: false });
}
