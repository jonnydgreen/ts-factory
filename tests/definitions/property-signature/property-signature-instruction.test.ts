import { Definition } from '../../../definitions/definitions.ts';
import { ts, tsm } from '../../../deps.ts';
import { buildNodeFromDefinition } from '../../../instructions/instructions.utils.ts';
import { assertSnapshot, blocks } from '../../../test.deps.ts';
import { createTestName, TestDefinition } from '../../test-utils.ts';

blocks.describe('Property Signature', () => {
  const definitions: TestDefinition<Definition>[] = [
    {
      name: createTestName('should process a basic Property Signature'),
      input: {
        kind: ts.SyntaxKind.PropertySignature,
        name: 'hello',
      },
    },
    {
      name: createTestName(
        'should process a Property Signature with an identifier as the name',
      ),
      input: {
        kind: ts.SyntaxKind.PropertySignature,
        name: { kind: ts.SyntaxKind.Identifier, text: 'hello' },
      },
    },
    {
      name: createTestName('should process a Property Signature with modifiers'),
      input: {
        kind: ts.SyntaxKind.PropertySignature,
        name: 'hello',
        modifiers: [{ kind: ts.SyntaxKind.ReadonlyKeyword }],
      },
    },
    {
      name: createTestName('should process a Property Signature with a type'),
      input: {
        kind: ts.SyntaxKind.PropertySignature,
        name: 'hello',
        type: { kind: ts.SyntaxKind.StringKeyword },
      },
    },
    {
      name: createTestName('should process a Property Signature with a question token'),
      input: {
        kind: ts.SyntaxKind.PropertySignature,
        name: 'hello',
        type: { kind: ts.SyntaxKind.StringKeyword },
        questionToken: { kind: ts.SyntaxKind.QuestionToken },
      },
    },
  ];

  for (const definition of definitions) {
    blocks.it({
      name: definition.name,
      fn: async (t) => {
        // Act
        const node = buildNodeFromDefinition(definition.input);

        // Assert
        await assertSnapshot(t, tsm.printNode(node));
      },
      ignore: definition.ignore,
      only: definition.only,
    });
  }
});
