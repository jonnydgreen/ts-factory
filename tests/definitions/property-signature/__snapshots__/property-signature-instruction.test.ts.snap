export const snapshot = {};

snapshot[`Property Signature > should process a basic Property Signature 1`] = `"hello;"`;

snapshot[`Property Signature > should process a Property Signature with an identifier as the name 1`] = `"hello;"`;

snapshot[`Property Signature > should process a Property Signature with modifiers 1`] = `"readonly hello;"`;

snapshot[`Property Signature > should process a Property Signature with a type 1`] = `"hello: string;"`;

snapshot[`Property Signature > should process a Property Signature with a question token 1`] = `"hello?: string;"`;
